galaxyproject/galaxy
# Expose input step order and rework calculation of node position in the workflow editor

Implements https://github.com/galaxyproject/galaxy/issues/13393 and some related issues:

- A step index is added to each of the steps within the workflow editor (see the screenshot), informing the user how they will be ordered in the workflow run form.
-  Step index depends on the distance of the node to the canvas origin, which is the top left of the window, but this is not fixed - it moves when the window is dragged around, so step index changes unexpectedly. This is fixed by normalizing the node coordinates by subtracting the minimum `top` and minimum `left` coordinates from the positions of each node when saving, so no node can have negative coordinates. It has the pleasant side effect that dragging the canvas around has no effect on the node positions (unlike the current behaviour).
- The normalization means that node positions are now effectively defined relative to each other, rather than in an absolute coordinate frame. The `update_step_position` refactor action was therefore updated so that the coordinates provided are used to shift the current position, rather than setting a new absolute position. Hopefully this is ok, I guess that action is not being used too widely atm.
- Previously all attributes returned by `getBoundingClientRect` (https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) were being saved, although only `left` and `top` should be needed. This has been changed so that for new workflows only `left` and `top` are saved.

![image](https://user-images.githubusercontent.com/32272674/160878487-32e1d00a-a64d-45a1-bff7-bf9496d7f41c.png)

## How to test the changes?
(Select all options that apply)
  1. Open a workflow, note the indices in front of the step labels (see the screenshot)
  2. Drag some steps around and save.
  3. Step order should have changed.

## License
